promql/engine.go

```diff
@@ -544,20 +543,36 @@ func (ng *Engine) populateSeries(ctx context.Context, q storage.Queryable, s *Ev
 				params.End = params.End - offsetMilliseconds
 			}
 
-			set, err = querier.Select(params, n.LabelMatchers...)
+			set, err, wrn = querier.Select(params, n.LabelMatchers...)
+			warnings = append(warnings, wrn...)
 			if err != nil {
 				level.Error(ng.logger).Log("msg", "error selecting series set", "err", err)
 				return err
 			}
-			n.series, err = expandSeriesSet(ctx, set)
-			if err != nil {
-				level.Error(ng.logger).Log("msg", "error expanding series set", "err", err)
-				return err
+			n.unexpandedSeriesSet = set
+		}
+		return nil
+	})
+	Inspect(s.Expr, func(node Node, path []Node) error {
+		switch n := node.(type) {
+		case *VectorSelector:
+			if n.unexpandedSeriesSet != nil {
```

Comment:
In what circumstances is this nil?

It could also be combined with the above, or you could go back to doing it at eval time (which is a bit more memory efficient).

Reply:
`n.unexpandedSeriesSet` would be nil any time that `querier.Select` returns an error. I put it here because we wanted to factor out the expansion. Unless I misunderstood, did you just mean to take that portion of code put it in a function instead of repeating it a few times?

Reply:
If querier.Select returns an error, the query has already failed.

> did you just mean to take that portion of code put it in a function instead of repeating it a few times?

Yes